client/transaction-pool/src/lib.rs

```diff
@@ -130,3 +175,158 @@ impl<PoolApi, Block> TransactionPool for BasicPool<PoolApi, Block>
+	Light(RevalidationStatus<N>)
+}
+
+impl<N: Clone + Copy + SimpleArithmetic> RevalidationStrategy<N> {
+	pub fn clear(&mut self) {
+		if let Self::Light(status) = self { status.clear() }
+	}
+
+	pub fn resubmit_required(&mut self) -> bool {
+		if let Self::Light(_) = self { return false } else { return true }
+	}
+
+	pub fn is_required(
+		&mut self,
+		block: N,
+		revalidate_time_period: Option<std::time::Duration>,
+		revalidate_block_period: Option<N>,
+	) -> bool {
+		if let Self::Light(status) = self {
+			status.is_required(block, revalidate_time_period, revalidate_block_period)
+		} else { true }
+	}
+
+	pub fn amount(&self) -> Option<usize> {
+		match self {
+			Self::Light(_) => None,
+			Self::Always => Some(16),
+		}
+	}
+}
+
+impl<N: Clone + Copy + SimpleArithmetic> RevalidationStatus<N> {
+	/// Called when revalidation is completed.
+	pub fn clear(&mut self) {
+		*self = Self::NotScheduled;
+	}
+
+	/// Returns true if revalidation is required.
+	pub fn is_required(
+		&mut self,
+		block: N,
+		revalidate_time_period: Option<std::time::Duration>,
+		revalidate_block_period: Option<N>,
+	) -> bool {
+		match *self {
+			Self::NotScheduled => {
+				*self = Self::Scheduled(
+					revalidate_time_period.map(|period| Instant::now() + period),
+					revalidate_block_period.map(|period| block + period),
+				);
+				false
+			},
+			Self::Scheduled(revalidate_at_time, revalidate_at_block) => {
+				let is_required = revalidate_at_time.map(|at| Instant::now() >= at).unwrap_or(false)
+					|| revalidate_at_block.map(|at| block >= at).unwrap_or(false);
+				if is_required {
+					*self = Self::InProgress;
+				}
+				is_required
+			},
+			Self::InProgress => false,
+		}
+	}
+}
+
+impl<PoolApi, Block> MaintainedTransactionPool for BasicPool<PoolApi, Block>
+where
+	Block: BlockT,
+	PoolApi: 'static + sc_transaction_graph::ChainApi<Block=Block, Hash=Block::Hash, Error=error::Error>,
+{
+	fn maintain(&self, id: &BlockId<Self::Block>, retracted: &[BlockHash<Self>]) -> Pin<Box<dyn Future<Output=()> + Send>> {
+		let id = id.clone();
+		let pool = self.pool.clone();
+		let api = self.api.clone();
+
+		let block_number = match api.block_id_to_number(&id) {
+			Ok(Some(number)) => number,
+			_ => {
+				log::trace!(target: "txqueue", "Skipping chain event - no numbrer for that block {:?}", id);
+				return Box::pin(ready(()));
+			}
+		};
+
+		let (is_revalidation_required, is_resubmit_required) = {
+			let mut lock = self.revalidation_strategy.lock();
+			(
+				lock.is_required(
+					block_number,
+					Some(std::time::Duration::from_secs(60)),
+					Some(20.into()),
+				),
+				lock.resubmit_required()
+			)
+		};
+
+		let revalidation_status = self.revalidation_strategy.clone();
+		let revalidation_amount = revalidation_status.lock().amount();
+		let retracted = retracted.to_vec();
+
+		async move {
+			let hashes = api.block_body(&id).await
+				.unwrap_or_else(|e| {
+					log::warn!("Prune known transactions: error request {:?}!", e);
```

Comment:
```suggestion
					log::error!(target: "txpool", "Unable to get block body for pruning: {:?}!", e);
```
as I don't think it should ever happen.

Reply:
It can, if light client fails the block body fetch, which can happen from time to time